File: macos/EngramCoreWrite/ProjectMove/JsonlPatch.swift
// macos/EngramCoreWrite/ProjectMove/JsonlPatch.swift
// Mirrors src/core/project-move/jsonl-patch.ts (Node parity baseline).
//
// Byte-level path rewriter for JSONL session files. Replaces every
// occurrence of `oldPath` with `newPath` where the match is followed by
// a path-terminator character (or end-of-input). Terminators: `"` `'` `/`
// `\` `<` `>` `]` `)` `}` backtick whitespace EOF. NOT terminators (so
// `.bak`, `-v2`, `_dir` survive): `.` `,` `;` `-` `_` and alphanumerics.
//
// **Code rule:** this module MUST NOT call JSONDecoder on the input. It
// works at byte level so Python mvp.py and Swift produce byte-identical
// output (diff-test invariant).
//
// Strict UTF-8 decode: `String(data:encoding:.utf8)` returns nil on
// malformed sequences instead of inserting U+FFFD, so the strict-decode
// contract from Node maps directly. Throw `InvalidUtf8Error` rather than
// corrupt the file's original bytes.
import Darwin
import Foundation

public struct InvalidUtf8Error: ProjectMoveError, Equatable {
    public let detail: String
    public init(detail: String) { self.detail = detail }
    public var errorName: String { "InvalidUtf8Error" }
    public var errorMessage: String { "patchBuffer: input is not valid UTF-8 (\(detail))" }
}

public struct ConcurrentModificationError: ProjectMoveError, Equatable {
    public let filePath: String
    public let oldMtime: Double
    public let newMtime: Double
    public init(filePath: String, oldMtime: Double, newMtime: Double) {
        self.filePath = filePath
        self.oldMtime = oldMtime
        self.newMtime = newMtime
    }
    public var errorName: String { "ConcurrentModificationError" }
    public var errorMessage: String {
        "patchFile: \(filePath) was modified during patch (mtime \(oldMtime) → \(newMtime)). " +
        "Another process wrote to the file between read and rename. " +
        "Safe fallback: retry later; orchestrator should retry with exponential backoff."
    }
}

public enum JsonlPatchError: Error, Equatable {
    case fileTooLarge(path: String, size: Int64, limit: Int64)
    case ioError(path: String, errno: Int32, message: String)
}

public struct PatchResult: Equatable {
    public let data: Data
    public let count: Int
    public init(data: Data, count: Int) {
        self.data = data
        self.count = count
    }
}

public enum JsonlPatch {
    /// Path-terminator lookahead. Excludes `.` `,` `;` `-` `_` and
    /// alphanumerics so paths like `/p/file.bak`, `/p-v2`, `/p_dir`
    /// don't false-positive.
    public static let pathTerminatorLookahead = #"(?=["'/\\<>\])}`\s]|$)"#

    /// Hard cap: refuse to patch JSONL above this size in memory. The
    /// orchestrator either streams or fails fast — current callers all
    /// fail fast.
    public static let maxInMemoryBytes: Int64 = 128 * 1024 * 1024

    /// Replace `oldPath` with `newPath` in `data`, preserving all other
    /// bytes. Returns a new `Data` value plus the number of replacements.
    /// Round-4 NFD fallback: if the UTF-8 source was decomposed (HFS+
    /// volume) but the user typed the rename target in NFC, retry the
    /// match against `oldPath.decomposedStringWithCanonicalMapping`.
    public static func patchBuffer(
        _ data: Data,
        oldPath: String,
        newPath: String
    ) throws -> PatchResult {
        if oldPath.isEmpty || oldPath == newPath {
            return PatchResult(data: data, count: 0)
        }
        guard let text = String(data: data, encoding: .utf8) else {
            throw InvalidUtf8Error(detail: "input bytes are not a valid UTF-8 sequence")
        }
        var totalCount = 0
        var working = replaceWithTerminator(
            in: text,
            needle: oldPath,
            replacement: newPath,
            count: &totalCount
        )
        let oldNfd = oldPath.decomposedStringWithCanonicalMapping
        if oldNfd != oldPath {
            working = replaceWithTerminator(
                in: working,
                needle: oldNfd,
                replacement: newPath,
                count: &totalCount
            )
        }
        if totalCount == 0 {
            return PatchResult(data: data, count: 0)
        }
        return PatchResult(data: Data(working.utf8), count: totalCount)
    }

    /// Byte-level literal replace of `<oldPath>."` → `<newPath>."`. Mirrors
    /// mvp.py's `auto_fix_dot_quote`: a `.` followed by `"` cannot be a
    /// filename extension (it's a sentence-end quote), so it's safe to
    /// rewrite even though the main regex excludes `.`.
    public static func autoFixDotQuote(
        _ data: Data,
        oldPath: String,
        newPath: String
    ) -> PatchResult {
        let needle = Data((oldPath + ".\"").utf8)
        let replacement = Data((newPath + ".\"").utf8)
        if needle == replacement || data.range(of: needle) == nil {
            return PatchResult(data: data, count: 0)
        }
        var output = Data()
        output.reserveCapacity(data.count)
        var cursor = data.startIndex
        var count = 0
        while cursor <= data.endIndex {
            if let hit = data.range(of: needle, in: cursor..<data.endIndex) {
                output.append(data[cursor..<hit.lowerBound])
                output.append(replacement)
                cursor = hit.upperBound
                count += 1
            } else {
                output.append(data[cursor..<data.endIndex])
                break
            }
        }
        if count == 0 {
            return PatchResult(data: data, count: 0)
        }
        return PatchResult(data: output, count: count)
    }

    /// Combined main + dot-quote sweep — use this in the orchestrator's
    /// CAS window so both transformations are atomic and reversible.
    public static func patchBufferWithDotQuote(
        _ data: Data,
        oldPath: String,
        newPath: String
    ) throws -> PatchResult {
        let first = try patchBuffer(data, oldPath: oldPath, newPath: newPath)
        let second = autoFixDotQuote(first.data, oldPath: oldPath, newPath: newPath)
        return PatchResult(data: second.data, count: first.count + second.count)
    }

    /// Read, patch, atomically rename a single file with mtime CAS. Returns
    /// the number of replacements (0 = file untouched, no write performed).
    /// Re-stats twice — once before the write, once after — so a concurrent
    /// writer between read and rename surfaces as `ConcurrentModificationError`.
    public static func patchFile(
        at filePath: String,
        oldPath: String,
        newPath: String
    ) throws -> Int {
        try rejectSymlinkSource(filePath)
        let attrsBefore = try FileManager.default.attributesOfItem(atPath: filePath)
        let sizeBefore = (attrsBefore[.size] as? NSNumber)?.int64Value ?? 0
        if sizeBefore > maxInMemoryBytes {
            return try patchFileStreaming(
                at: filePath,
                oldPath: oldPath,
                newPath: newPath,
                attrsBefore: attrsBefore
            )
        }
        let before = try snapshot(path: filePath)

        let buf = try Data(contentsOf: URL(fileURLWithPath: filePath))
        let res = try patchBufferWithDotQuote(buf, oldPath: oldPath, newPath: newPath)
        if res.count == 0 { return 0 }

        // First CAS: file must not have moved between read and now.
        let after = try snapshot(path: filePath)
        if after != before {
            throw ConcurrentModificationError(
                filePath: filePath,
                oldMtime: Double(before.mtimeSec) * 1000 + Double(before.mtimeNsec) / 1_000_000,
                newMtime: Double(after.mtimeSec) * 1000 + Double(after.mtimeNsec) / 1_000_000
            )
        }

        let tmpPath = "\(filePath).engram-tmp-\(getpid())-\(randomToken())"
        do {
            try res.data.write(to: URL(fileURLWithPath: tmpPath))
            if let permissions = attrsBefore[.posixPermissions] as? NSNumber {
                chmod(tmpPath, mode_t(permissions.intValue))
            }
            try fsyncFile(at: tmpPath)
        } catch {
            _ = try? FileManager.default.removeItem(atPath: tmpPath)
            throw JsonlPatchError.ioError(
                path: tmpPath, errno: errno, message: error.localizedDescription
            )
        }

        // Second CAS: closes the gap between the first stat and rename.
        let final = try snapshot(path: filePath)
        if final != before {
            _ = try? FileManager.default.removeItem(atPath: tmpPath)
            throw ConcurrentModificationError(
                filePath: filePath,
                oldMtime: Double(before.mtimeSec) * 1000 + Double(before.mtimeNsec) / 1_000_000,
                newMtime: Double(final.mtimeSec) * 1000 + Double(final.mtimeNsec) / 1_000_000
            )
        }

        if Darwin.rename(tmpPath, filePath) != 0 {
            let code = errno
            _ = try? FileManager.default.removeItem(atPath: tmpPath)
            throw JsonlPatchError.ioError(
                path: filePath, errno: code, message: String(cString: strerror(code))
            )
        }
        fsyncDirectory(for: filePath)
        return res.count
    }

    // MARK: - internals

    private struct FileSnapshot: Equatable {
        let device: UInt64
        let inode: UInt64
        let size: Int64
        let mtimeSec: Int
        let mtimeNsec: Int
    }

    private static func snapshot(path: String) throws -> FileSnapshot {
        var info = stat()
        if lstat(path, &info) != 0 {
            throw JsonlPatchError.ioError(
                path: path,
                errno: errno,
                message: String(cString: strerror(errno))
            )
        }
        return FileSnapshot(
            device: UInt64(info.st_dev),
            inode: UInt64(info.st_ino),
            size: Int64(info.st_size),
            mtimeSec: Int(info.st_mtimespec.tv_sec),
            mtimeNsec: Int(info.st_mtimespec.tv_nsec)
        )
    }

    private static func rejectSymlinkSource(_ filePath: String) throws {
        var info = stat()
        if lstat(filePath, &info) != 0 {
            throw JsonlPatchError.ioError(
                path: filePath,
                errno: errno,
                message: String(cString: strerror(errno))
            )
        }
        if (info.st_mode & S_IFMT) == S_IFLNK {
            throw JsonlPatchError.ioError(
                path: filePath,
                errno: ELOOP,
                message: "source is symlink"
            )
        }
    }

    private static func patchFileStreaming(
        at filePath: String,
        oldPath: String,
        newPath: String,
        attrsBefore: [FileAttributeKey: Any]
    ) throws -> Int {
        let before = try snapshot(path: filePath)
        let tmpPath = "\(filePath).engram-tmp-\(getpid())-\(randomToken())"
        FileManager.default.createFile(atPath: tmpPath, contents: nil)
        if let permissions = attrsBefore[.posixPermissions] as? NSNumber {
            chmod(tmpPath, mode_t(permissions.intValue))
        }

        let input = try FileHandle(forReadingFrom: URL(fileURLWithPath: filePath))
        let output = try FileHandle(forWritingTo: URL(fileURLWithPath: tmpPath))
        var total = 0
        var carry = Data()
        let carryLimit = max(
            oldPath.lengthOfBytes(using: .utf8),
            oldPath.decomposedStringWithCanonicalMapping.lengthOfBytes(using: .utf8)
        ) + 8

        do {
            while true {
                let chunk = try input.read(upToCount: 1024 * 1024) ?? Data()
                if chunk.isEmpty { break }
                var combined = Data()
                combined.reserveCapacity(carry.count + chunk.count)
                combined.append(carry)
                combined.append(chunk)

                guard combined.count > carryLimit else {
                    carry = combined
                    continue
                }
                var processCount = combined.count - carryLimit
                var segment = combined.prefix(processCount)
                var decoded = String(data: segment, encoding: .utf8)
                var attempts = 0
                while decoded == nil && processCount > 0 && attempts < 4 {
                    processCount -= 1
                    attempts += 1
                    segment = combined.prefix(processCount)
                    decoded = String(data: segment, encoding: .utf8)
                }
                guard let decoded else {
                    throw InvalidUtf8Error(detail: "input bytes are not a valid UTF-8 sequence")
                }
                let patched = try patchBufferWithDotQuote(
                    Data(decoded.utf8),
                    oldPath: oldPath,
                    newPath: newPath
                )
                total += patched.count
                try output.write(contentsOf: patched.data)
                carry = combined.suffix(combined.count - processCount)
            }

            let tail = try patchBufferWithDotQuote(
                carry,
                oldPath: oldPath,
                newPath: newPath
            )
            total += tail.count
            try output.write(contentsOf: tail.data)
            output.synchronizeFile()
            try output.close()
            try input.close()

            if total == 0 {
                _ = try? FileManager.default.removeItem(atPath: tmpPath)
                return 0
            }

            let after = try snapshot(path: filePath)
            if after != before {
                _ = try? FileManager.default.removeItem(atPath: tmpPath)
                throw ConcurrentModificationError(
                    filePath: filePath,
                    oldMtime: Double(before.mtimeSec) * 1000 + Double(before.mtimeNsec) / 1_000_000,
                    newMtime: Double(after.mtimeSec) * 1000 + Double(after.mtimeNsec) / 1_000_000
                )
            }
            if Darwin.rename(tmpPath, filePath) != 0 {
                let code = errno
                _ = try? FileManager.default.removeItem(atPath: tmpPath)
                throw JsonlPatchError.ioError(
                    path: filePath,
                    errno: code,
                    message: String(cString: strerror(code))
                )
            }
            fsyncDirectory(for: filePath)
            return total
        } catch {
            try? output.close()
            try? input.close()
            _ = try? FileManager.default.removeItem(atPath: tmpPath)
            throw error
        }
    }

    private static func fsyncDirectory(for filePath: String) {
        let directory = (filePath as NSString).deletingLastPathComponent
        let fd = Darwin.open(directory, O_RDONLY)
        guard fd >= 0 else { return }
        _ = Darwin.fsync(fd)
        Darwin.close(fd)
    }

    private static func fsyncFile(at path: String) throws {
        let fd = Darwin.open(path, O_RDONLY)
        if fd < 0 {
            throw JsonlPatchError.ioError(
                path: path,
                errno: errno,
                message: String(cString: strerror(errno))
            )
        }
        defer { Darwin.close(fd) }
        if Darwin.fsync(fd) != 0 {
            throw JsonlPatchError.ioError(
                path: path,
                errno: errno,
                message: String(cString: strerror(errno))
            )
        }
    }

    private static func replaceWithTerminator(
        in text: String,
        needle: String,
        replacement: String,
        count: inout Int
    ) -> String {
        let escaped = NSRegularExpression.escapedPattern(for: needle)
        let pattern = escaped + pathTerminatorLookahead
        guard let regex = try? NSRegularExpression(pattern: pattern, options: []) else {
            return text
        }
        let nsText = text as NSString
        let range = NSRange(location: 0, length: nsText.length)
        let matches = regex.matches(in: text, options: [], range: range)
        if matches.isEmpty {
            return text
        }
        var result = text
        for match in matches.reversed() {
            if let r = Range(match.range, in: result) {
                result.replaceSubrange(r, with: replacement)
            }
        }
        count += matches.count
        return result
    }

    private static func mtimeMs(_ attrs: [FileAttributeKey: Any]) -> Double {
        guard let date = attrs[.modificationDate] as? Date else { return 0 }
        return date.timeIntervalSince1970 * 1000
    }

    private static func randomToken() -> String {
        var raw = [UInt8](repeating: 0, count: 4)
        arc4random_buf(&raw, 4)
        return raw.map { String(format: "%02x", $0) }.joined()
    }
}

File: src/core/project-move/jsonl-patch.ts
// src/core/project-move/jsonl-patch.ts — byte-level path rewriter for JSONL
//
// 1:1 translation of mvp.py:patch_file. Semantics:
//   - Replace every occurrence of `oldPath` with `newPath` where the match
//     is followed by a path-terminator character (or end-of-input).
//   - Terminators: " ' / \ < > ] ) } backtick whitespace EOF
//   - NOT terminators (by design, to protect /path/name.bak, /path/name-v2,
//     /path/name_suffix from being truncated): . , ; - _ and alphanumerics.
//
// **Code rule (Gemini/Codex review):** this module MUST NOT call JSON.parse
// on the input. It operates at byte-level through Buffer/UTF-8 string round-
// trip so Python mvp and TS produce byte-identical output (diff-test invariant).

import { createReadStream } from 'node:fs';
import {
  chmod,
  open,
  readFile,
  rename,
  stat,
  unlink,
  writeFile,
} from 'node:fs/promises';
import { dirname } from 'node:path';

/**
 * Strict UTF-8 decoder — throws on malformed sequences instead of silently
 * replacing with U+FFFD. Required for byte-parity with Python mvp.py:
 * if we can't be byte-identical, we must refuse to write rather than
 * corrupt the file's original bytes.
 */
const STRICT_UTF8_DECODER = new TextDecoder('utf-8', { fatal: true });

export class InvalidUtf8Error extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'InvalidUtf8Error';
  }
}

/**
 * Path-terminator lookahead — regex literal (not constructed from strings)
 * to avoid double-escape hazards. Matches one of:
 *   "  '  /  \  <  >  ]  )  }  backtick  whitespace  end-of-input
 *
 * Excludes: . , ; - _ and alphanumerics (so `/path/file.bak` / `/path-v2`
 * / `/path_dir` don't false-positive).
 */
const PATH_TERMINATOR_LOOKAHEAD = /(?=["'/\\<>\])}`\s]|$)/;

/** Escape regex metacharacters so `oldPath` is matched literally. */
function escapeRegex(literal: string): string {
  return literal.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

/**
 * Build the path-rewrite regex for a given old path. Keep it local to
 * each patch call (patchBuffer is pure; no global state).
 */
function buildRegex(oldPath: string): RegExp {
  return new RegExp(
    escapeRegex(oldPath) + PATH_TERMINATOR_LOOKAHEAD.source,
    'g',
  );
}

export interface PatchResult {
  buffer: Buffer;
  count: number;
}

/**
 * Replace `oldPath` with `newPath` in the buffer, preserving all other bytes.
 * Returns a new buffer + the number of replacements made.
 *
 * mvp.py equivalent: `patch_file(file, old, new)` body excluding the file I/O.
 */
export function patchBuffer(
  data: Buffer,
  oldPath: string,
  newPath: string,
): PatchResult {
  if (oldPath === '' || oldPath === newPath) {
    return { buffer: data, count: 0 };
  }
  // Strict decode — invalid UTF-8 bytes would be silently mangled into
  // U+FFFD by Buffer.toString, breaking byte-parity with Python mvp.py.
  // Refuse to touch the file rather than corrupt it.
  let text: string;
  try {
    text = STRICT_UTF8_DECODER.decode(data);
  } catch (err) {
    throw new InvalidUtf8Error(
      `patchBuffer: input is not valid UTF-8 (${(err as Error).message})`,
    );
  }
  const { text: out, count } = patchText(text, oldPath, newPath);
  if (count === 0) return { buffer: data, count: 0 };
  return { buffer: Buffer.from(out, 'utf8'), count };
}

function patchText(
  text: string,
  oldPath: string,
  newPath: string,
): { text: string; count: number } {
  const primary = buildRegex(oldPath);
  let count = 0;
  let out = text.replace(primary, () => {
    count++;
    return newPath;
  });
  // Round 4 (Codex #3): macOS HFS+ stores filenames in NFD; an AI CLI
  // running on such a volume may also write the *path* in NFD into its
  // JSONL. If the user typed the rename target in NFC (common), the
  // primary regex will miss those occurrences. Retry with an NFD-form
  // needle as a fallback — the replacement stays NFC so the FS wins
  // the tie (consumers open files by NFC path anyway).
  const oldNfd = oldPath.normalize('NFD');
  if (oldNfd !== oldPath) {
    const secondary = buildRegex(oldNfd);
    out = out.replace(secondary, () => {
      count++;
      return newPath;
    });
  }
  return { text: out, count };
}

function patchTextWithDotQuote(
  text: string,
  oldPath: string,
  newPath: string,
): { text: string; count: number } {
  const first = patchText(text, oldPath, newPath);
  const needle = `${oldPath}."`;
  if (!first.text.includes(needle)) return first;
  let count = first.count;
  const out = first.text.split(needle).join(`${newPath}."`);
  count += first.text.split(needle).length - 1;
  return { text: out, count };
}

/**
 * Replace `<oldPath>."` with `<newPath>."` — mvp.py:auto_fix_dot_quote.
 *
 * Rationale: the main regex excludes `.` from terminators to preserve
 * `.bak` / `.py` paths. But a `.` followed by `"` cannot be a filename
 * extension — it's a sentence-end quote. Safe to fix after the fact.
 */
export function autoFixDotQuote(
  data: Buffer,
  oldPath: string,
  newPath: string,
): PatchResult {
  const needle = Buffer.from(`${oldPath}."`, 'utf8');
  const replacement = Buffer.from(`${newPath}."`, 'utf8');
  if (data.indexOf(needle) === -1) return { buffer: data, count: 0 };

  // Use split/join on the buffer representation via bytes. Node's Buffer
  // doesn't have replaceAll, so we manually build.
  const parts: Buffer[] = [];
  let cursor = 0;
  let count = 0;
  while (cursor <= data.length) {
    const hit = data.indexOf(needle, cursor);
    if (hit === -1) {
      parts.push(data.subarray(cursor));
      break;
    }
    parts.push(data.subarray(cursor, hit));
    parts.push(replacement);
    cursor = hit + needle.length;
    count++;
  }
  return { buffer: Buffer.concat(parts), count };
}

/** Max size we'll patch in-memory before switching to streaming I/O. */
const MAX_IN_MEMORY_BYTES = 128 * 1024 * 1024; // 128 MiB

export class ConcurrentModificationError extends Error {
  constructor(filePath: string, oldMtime: number, newMtime: number) {
    super(
      `patchFile: ${filePath} was modified during patch (mtime ${oldMtime} → ${newMtime}). ` +
        'Another process wrote to the file between read and rename. ' +
        'Safe fallback: retry later; orchestrator should retry with exponential backoff.',
    );
    this.name = 'ConcurrentModificationError';
  }
}

/**
 * Combined path-rewrite: main regex (path-terminator aware) + dot-quote
 * fallback. Round 4 Critical: previously the orchestrator ran the dot-quote
 * sweep as a SEPARATE read-write pass bypassing CAS, which could overwrite
 * concurrent writes. Folding it into patchBuffer keeps the transformation
 * atomic AND makes it automatically reversible by compensation (since
 * compensation replays patchFile with src/dst swapped).
 */
export function patchBufferWithDotQuote(
  data: Buffer,
  oldPath: string,
  newPath: string,
): PatchResult {
  const first = patchBuffer(data, oldPath, newPath);
  // autoFixDotQuote still works on the remaining pattern — after the main
  // regex, any `<oldPath>."` that slipped through (path-terminator regex
  // excludes `.` to preserve `.bak`) gets caught here.
  const second = autoFixDotQuote(first.buffer, oldPath, newPath);
  return {
    buffer: second.buffer,
    count: first.count + second.count,
  };
}

/**
 * Read, patch, write a single file atomically with compare-and-swap mtime
 * protection. Uses `<file>.engram-tmp-<pid>-<rand>` + rename.
 *
 * Concurrency protocol (Gemini blocker #2): Captures `mtimeMs` at read time,
 * re-stats immediately before rename. If they differ, another process wrote
 * to the file during our patch — we abort rather than silently overwriting
 * (and losing) the concurrent writer's appended data.
 *
 * The caller (orchestrator) should retry on `ConcurrentModificationError`
 * with a short backoff, OR acquire the global `.project-move.lock` and
 * advise users not to have AI CLIs open on the affected project.
 *
 * Returns the number of replacements (0 = not touched, no write, no CAS check).
 * mvp.py equivalent: `patch_file(file, old, new)` including I/O (but mvp.py
 * has no CAS — it silently clobbers; this is the improvement).
 *
 * Round 4: uses `patchBufferWithDotQuote` so the dot-quote fallback sweep
 * happens within the same CAS window as the main rewrite — previously the
 * orchestrator ran the dot-quote pass separately and bypassed CAS.
 */
export async function patchFile(
  filePath: string,
  oldPath: string,
  newPath: string,
): Promise<number> {
  const stBefore = await stat(filePath);
  const before = snapshot(stBefore);
  if (stBefore.size > MAX_IN_MEMORY_BYTES) {
    return await patchFileStreaming(filePath, oldPath, newPath, stBefore);
  }

  const buf = await readFile(filePath);
  const res = patchBufferWithDotQuote(buf, oldPath, newPath);
  if (res.count === 0) return 0;

  // Re-stat BEFORE we write — if the file changed since we read it, bail out.
  assertUnchanged(filePath, before, await stat(filePath));

  const tmp = tempPath(filePath);
  await writeFile(tmp, res.buffer);
  await chmod(tmp, stBefore.mode & 0o777).catch(() => {});

  // Final CAS — one more stat right before rename to close the gap between
  // the check above and the rename below.
  try {
    assertUnchanged(filePath, before, await stat(filePath));
  } catch (err) {
    await unlink(tmp).catch(() => {});
    throw err;
  }
  await rename(tmp, filePath);
  await fsyncPath(dirname(filePath)).catch(() => {});
  return res.count;
}

type FileSnapshot = {
  dev: number;
  ino: number;
  size: number;
  mtimeMs: number;
};

function snapshot(st: Awaited<ReturnType<typeof stat>>): FileSnapshot {
  return {
    dev: Number(st.dev),
    ino: Number(st.ino),
    size: Number(st.size),
    mtimeMs: Number(st.mtimeMs),
  };
}

function assertUnchanged(
  filePath: string,
  before: FileSnapshot,
  after: Awaited<ReturnType<typeof stat>>,
) {
  const next = snapshot(after);
  if (
    next.dev !== before.dev ||
    next.ino !== before.ino ||
    next.size !== before.size ||
    next.mtimeMs !== before.mtimeMs
  ) {
    throw new ConcurrentModificationError(
      filePath,
      before.mtimeMs,
      next.mtimeMs,
    );
  }
}

function tempPath(filePath: string): string {
  return `${filePath}.engram-tmp-${process.pid}-${Math.random()
    .toString(36)
    .slice(2, 8)}`;
}

function streamingCarryChars(oldPath: string): number {
  return Math.max(oldPath.length, oldPath.normalize('NFD').length) + 4;
}

async function fsyncPath(path: string) {
  const handle = await open(path, 'r');
  try {
    await handle.sync();
  } finally {
    await handle.close();
  }
}

async function patchFileStreaming(
  filePath: string,
  oldPath: string,
  newPath: string,
  stBefore: Awaited<ReturnType<typeof stat>>,
): Promise<number> {
  const before = snapshot(stBefore);
  const tmp = tempPath(filePath);
  const decoder = new TextDecoder('utf-8', { fatal: true });
  const carryChars = streamingCarryChars(oldPath);
  let carry = '';
  let total = 0;
  let out: Awaited<ReturnType<typeof open>> | undefined;
  const mode = Number(stBefore.mode) & 0o777;

  try {
    out = await open(tmp, 'wx', mode);
    await chmod(tmp, mode).catch(() => {});

    for await (const chunk of createReadStream(filePath, {
      highWaterMark: 1024 * 1024,
    })) {
      let decoded: string;
      try {
        decoded = decoder.decode(chunk as Buffer, { stream: true });
      } catch (err) {
        throw new InvalidUtf8Error(
          `patchFile: input is not valid UTF-8 (${(err as Error).message})`,
        );
      }
      const working = carry + decoded;
      if (working.length <= carryChars) {
        carry = working;
        continue;
      }
      const processText = working.slice(0, working.length - carryChars);
      carry = working.slice(working.length - carryChars);
      const patched = patchTextWithDotQuote(processText, oldPath, newPath);
      total += patched.count;
      await out.write(Buffer.from(patched.text, 'utf8'));
    }

    const finalText = decoder.decode();
    const patchedTail = patchTextWithDotQuote(
      carry + finalText,
      oldPath,
      newPath,
    );
    total += patchedTail.count;
    await out.write(Buffer.from(patchedTail.text, 'utf8'));
    await out.sync();
    await out.close();
    out = undefined;

    if (total === 0) {
      await unlink(tmp).catch(() => {});
      return 0;
    }

    assertUnchanged(filePath, before, await stat(filePath));
    await rename(tmp, filePath);
    await fsyncPath(dirname(filePath)).catch(() => {});
    return total;
  } catch (err) {
    await out?.close().catch(() => {});
    await unlink(tmp).catch(() => {});
    throw err;
  }
}
